### PR TITLE
Introduce RPC Worker

### PR DESCRIPTION
**Pull request by skunert**
This PR is the next step towards https://github.com/paritytech/cumulus/issues/989.

Currently the collator opens a new RPC subscription every time we request a notification stream from the `RelayChainRPCInterface`. 

After this PR is merged, we will have only three open notification subscriptionsvia RPC:
- finalized block headers
- imported block headers
- new best block headers

A worker is introduced that will poll these three subscriptions regularly and distribute the contained headers to all listening streams. Requesting a stream from the `RelayChainRPCInterface` will return a new channel and inform the worker.
Since the streams are only used in the collator, and we have complete control, we use bounded channels with a limit of 20 items.

### CODE REVIEW

**Comment by davxy** [comment]
Really silly nit. Maybe we can stick to the notation where also the acronyms are usually written in camel case.
In other words `RPC` -> `Rpc`. Thus `RelayChainRPCClient` -> `RelayChainRpcClient`

Feel free to ignore me :-D

File: client/relay-chain-rpc-interface/src/rpc_client.rs [hunk 1]
```diff
@@ -0,0 +1,454 @@
...
+	best_header_listeners: Vec<Sender<PHeader>>,
+
+	// Incoming notification subscriptions
+	rpc_imported_header_subscription: Subscription<PHeader>,
+	rpc_finalized_header_subscription: Subscription<PHeader>,
+	rpc_best_header_subscription: Subscription<PHeader>,
+}
+
+/// Entry point to create [`RelayChainRPCClient`] and [`RPCStreamWorker`];
+pub async fn create_worker_client(
```
[the remark this answers is not in the document]
Comment (skunert) [reply client/relay-chain-rpc-interface/src/rpc_client.rs:91]: Yes, it is expected that users of this method spawn the worker task immediately. Otherwise the streams will not work and all usages inside the collator need the streams to work. I will look into spawning directly.

As for the users of this, there will be other usage soon. Currently we immediately wrap the client in a `RelayChainRPCInterface`. In the future we will pass the client into the relay chain subsystems additionally.

File: client/relay-chain-rpc-interface/src/rpc_client.rs [hunk 2]
```diff
@@ -0,0 +1,454 @@
...
+		retry_notify(
+			self.retry_strategy.clone(),
+			|| async {
+				self.ws_client.request(method, params.clone()).await.map_err(|err| match err {
+					JsonRpseeError::Transport(_) =>
+						backoff::Error::Transient { err, retry_after: None },
+					_ => backoff::Error::Permanent(err),
+				})
+			},
+			|error, dur| tracing::trace!(target: LOG_TARGET, %error, ?dur, "Encountered transport error, retrying."),
```
Comment (koute) [inline client/relay-chain-rpc-interface/src/rpc_client.rs:287]: Do we expect that this will trigger often? It should only trigger in cases like e.g. temporary disconnection, right? If so it seems to me that this is more of a `debug` log rather than a `trace` log. (But it's up to you; feel free to ignore this comment.)

File: client/relay-chain-rpc-interface/src/rpc_client.rs [hunk 3]
```diff
@@ -0,0 +1,454 @@
...
+fn handle_event_distribution(
+	event: Option<Result<PHeader, JsonRpseeError>>,
+	senders: &mut Vec<Sender<PHeader>>,
+) -> Result<(), String> {
+	match event {
+		// If sending fails, we remove the sender from the list because the receiver was dropped
+		Some(Ok(header)) => {
+			senders.retain_mut(|e| e.try_send(header.clone()).is_ok());
+			Ok(())
+		},
```
Comment (koute) [inline client/relay-chain-rpc-interface/src/rpc_client.rs:114]: The `try_send` can also return an error when the sender's channel is full, not only when it was dropped. (Since it's a bounded channel.)

File: client/relay-chain-rpc-interface/src/rpc_client.rs [hunk 4]
```diff
@@ -0,0 +1,454 @@
...
+pub enum NotificationRegisterMessage {
+	RegisterBestHeadListener(Sender<PHeader>),
+	RegisterImportListener(Sender<PHeader>),
+	RegisterFinalizationListener(Sender<PHeader>),
+}
+
+/// Worker that should be used in combination with [`RelayChainRPCClient`]. Must be polled to distribute header notifications to listeners.
+pub struct RPCStreamWorker {
+	// Communication channel with the RPC client
+	client_receiver: TracingUnboundedReceiver<NotificationRegisterMessage>,
```
[the remark this answers is not in the document]
Comment (skunert) [reply client/relay-chain-rpc-interface/src/rpc_client.rs:78]: I initially wanted to make it bounded, but the bounded sending requires `&mut self` which makes it super inconvenient in this case. Every stream getting method in the collator would need to take the mutable reference.

The channel is only used by the clients to request new notification streams, which are bounded. Also, the worker is started automatically and drains this channel when the client is constructed. The overall risk seems manageable to me in this case.

File: client/relay-chain-rpc-interface/src/rpc_client.rs (hunk 2 above)
Comment (skunert) [reply client/relay-chain-rpc-interface/src/rpc_client.rs:304]: I just moved this client without modification for this PR. Since we consider all of the RPC collation stuff experimental and recommend running relay and para node on the same machine, I do not see this being triggered. 

If there were any severe networking problems, the current RPC collator node would crash anyway, as we have no way to deal with broken subscriptions. So I consider the whole error handling an area to be overhauled down the road.

File: client/relay-chain-rpc-interface/src/rpc_client.rs (hunk 3 above)
Comment (skunert) [reply client/relay-chain-rpc-interface/src/rpc_client.rs]: I added a error print and do not remove the sender for now. Alternative would be to make the task fail.

File: client/relay-chain-rpc-interface/src/lib.rs [hunk 5]
```diff
@@ -26,274 +25,35 @@ use cumulus_primitives_core::{
...
-
-	async fn parachain_host_session_index_for_child(
-		&self,
-		at: PHash,
-	) -> Result<SessionIndex, RelayChainError> {
-		self.call_remote_runtime_function("ParachainHost_session_index_for_child", at, None::<()>)
-			.await
-	}
+mod rpc_client;
+pub use rpc_client::{create_client_and_start_worker, RelayChainRpcClient, RpcStreamWorker};
```
Comment (davxy) [inline client/relay-chain-rpc-interface/src/lib.rs:38]:
```suggestion
pub use rpc_client::{create_client_and_start_worker, RelayChainRpcClient};
```

File: client/relay-chain-rpc-interface/src/rpc_client.rs [hunk 6]
```diff
@@ -0,0 +1,471 @@
...
+/// Worker messages to register new notification listeners
+#[derive(Clone, Debug)]
+pub enum NotificationRegisterMessage {
+	RegisterBestHeadListener(Sender<PHeader>),
+	RegisterImportListener(Sender<PHeader>),
+	RegisterFinalizationListener(Sender<PHeader>),
+}
+
+/// Worker that should be used in combination with [`RelayChainRpcClient`]. Must be polled to distribute header notifications to listeners.
+pub struct RpcStreamWorker {
```
Comment (davxy) [inline client/relay-chain-rpc-interface/src/rpc_client.rs:76]:
```suggestion
struct RpcStreamWorker {
```

Because it is not directly exposed anymore

**Review by davxy — COMMENTED** [review]
LGTM, I've just added a suggestion to set the `RpcStreamWorker` as private

**Review by davxy — APPROVED** [review]

File: client/relay-chain-rpc-interface/src/rpc_client.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (koute) [reply client/relay-chain-rpc-interface/src/rpc_client.rs:78]: > I initially wanted to make it bounded, but the bounded sending requires `&mut self` which makes it super inconvenient in this case. Every stream getting method in the collator would need to take the mutable reference.

FYI, you could always just use [the channel from `tokio`](https://docs.rs/tokio/latest/tokio/sync/mpsc/struct.Sender.html) which doesn't require a `&mut self` to send. (:

File: client/relay-chain-rpc-interface/src/rpc_client.rs (hunk 3 above)
Comment (koute) [reply client/relay-chain-rpc-interface/src/rpc_client.rs]: Yeah, I think we need to think about what we should do if that happens and handle it appropriately, but considering this is still experimental and a work in progress I'm fine with it being as-is. But we'll need to revisit this issue before actually pushing this into production. Could you at least put like a TODO here as a reminder?

File: client/relay-chain-rpc-interface/src/rpc_client.rs (hunk 2 above)
Comment (koute) [reply client/relay-chain-rpc-interface/src/rpc_client.rs:304]: Fair enough!

**Review by koute — APPROVED** [review]
The unbounded channel's still making me a little nervous, but I'll leave the decision as to what to do with it up to you. Otherwise LGTM.

File: client/relay-chain-rpc-interface/src/rpc_client.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (skunert) [reply client/relay-chain-rpc-interface/src/rpc_client.rs]: Very good suggestion, was not aware of this difference. Switching to tokio then to be safe.

[the remark this answers is not in the document]
Comment (bkchr) [reply client/relay-chain-rpc-interface/src/rpc_client.rs]: Sorry for being late to the party. Not sure @koute why the usage of `unbounded` for the service to worker communication makes you nervous. I mean, that some is some well defined communication. We also don't want to drop any of this communication. We also don't need any backpressure here, otherwise @skunert could also just make the methods take `&mut self`. 

TLDR: I don't see any reason why we should not use unbounded here. If the worker is dropped the channel is closed and if we don't poll the unbounded receiver anymore, there is something broken anyway.

File: client/relay-chain-rpc-interface/src/rpc_client.rs [hunk 7]
```diff
@@ -0,0 +1,478 @@
...
+
+	fn send_register_message_to_worker(
+		&self,
+		message: NotificationRegisterMessage,
+	) -> Result<(), RelayChainError> {
+		match &self.to_worker_channel {
+			Some(channel) => {
+				channel
+					.try_send(message)
+					.map_err(|e| RelayChainError::WorkerCommunicationError(e.to_string()))?;
```
Comment (bkchr) [inline client/relay-chain-rpc-interface/src/rpc_client.rs:426]: We should not really fail because of service to worker communication channel is full.

File: client/relay-chain-rpc-interface/src/rpc_client.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (skunert) [reply client/relay-chain-rpc-interface/src/rpc_client.rs]: I think we should just default to bounded.

I just picked an unbounded one because the API was better for my usecase. The tokio channel solved this.

File: client/relay-chain-rpc-interface/src/rpc_client.rs (hunk 7 above)
Comment (skunert) [reply client/relay-chain-rpc-interface/src/rpc_client.rs:426]: I want it to fail for now, since the worker should really be fast enough to drain this channel. It is only used to request another notification stream.

If it runs full, I need to fix it. Once this is used in production we can make it more defensive. As mentioned, there will be another error handling pass anyway.

File: client/relay-chain-rpc-interface/src/rpc_client.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (bkchr) [reply client/relay-chain-rpc-interface/src/rpc_client.rs]: Bounded max sense when you need back pressure. Here you don't need any back pressure.

File: client/relay-chain-rpc-interface/src/rpc_client.rs [hunk 8]
```diff
@@ -0,0 +1,478 @@
...
+	task_manager: &mut TaskManager,
+) -> RelayChainResult<RelayChainRpcClient> {
+	let mut client = RelayChainRpcClient::new(url).await?;
+	let best_head_stream = client.subscribe_new_best_heads().await?;
+	let finalized_head_stream = client.subscribe_finalized_heads().await?;
+	let imported_head_stream = client.subscribe_imported_heads().await?;
+
+	let (worker, sender) =
+		RpcStreamWorker::new(imported_head_stream, best_head_stream, finalized_head_stream);
+	client.set_worker_channel(sender);
```
Comment (bkchr) [inline client/relay-chain-rpc-interface/src/rpc_client.rs:105]: Why don't we first create the worker and then the client? Then we don't need to make the `sender` an `Option`.

File: client/relay-chain-rpc-interface/src/rpc_client.rs (hunk 7 above)
Comment (bkchr) [reply client/relay-chain-rpc-interface/src/rpc_client.rs:426]: I get that you want to improve this later. However, the entire system here with worker/client is build in a way that both are working together in a tight way. Meaning, if the worker or the client stops the other end should also stop. Then both are also not required to have any kind of back pressure. You want that anything you send from the client side gets forwarded to the worker. Dropping here any data doesn't make any sense.

File: client/relay-chain-rpc-interface/src/rpc_client.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (koute) [reply client/relay-chain-rpc-interface/src/rpc_client.rs]: Sorry, it's just that I've seen enough code already that used an unbounded channels and was also well defined and looked innocuous but ended up causing problems. And a failure mode where it fails explicitly is easier to debug *if* something were to be broken here. Which is why if we'd end up in the "something's broken anyway" situation I'd prefer to have it alert us to that fact, instead of potentially just silently chugging along in the worst case.

This is just a "just in case" measure; as you say, yeah, we don't strictly *need* this here, but I also don't see why not to do this; in normal circumstances when everything works as it should (that is - the messages are handled fast enough that no backpressure will ever be triggered) it should essentially be functionally equivalent to an unbound channel.

That said, it's not a huge deal, so if you feel strongly about not having a bounded channel here I'm fine with reverting it. :P

[the remark this answers is not in the document]
Comment (bkchr) [reply client/relay-chain-rpc-interface/src/rpc_client.rs]: Okay okay, I "give" up 😂 but we should at least increase the channel size to 1000. If something is broken, we still will see it.

[the remark this answers is not in the document]
Comment (bkchr) [reply client/relay-chain-rpc-interface/src/rpc_client.rs]: I mean it probably makes no difference, but we use almost everywhere in our code the futures channel and futures select. That is something I don't want to die on, but yeah.

[the remark this answers is not in the document]
Comment (koute) [reply client/relay-chain-rpc-interface/src/rpc_client.rs]: Indeed. But I don't really see us ever dropping the `tokio` dependency, and if it's an integral part of `tokio` we might as well use it *if* needed. It's not like it's some random unmaintained crate fresh off crates.io.

I even once tried to compare whether switching from the futures' channel to tokio's channel would make a difference performance-wise (for a few particularly busy channels in substrate), but in that particular case I found the difference to be negligible. Or in other words, performance-wise they're comparable, so it shouldn't really matter which one we use *in general* (not that this aspect matters here; it doesn't), except `tokio`'s channels *do* have a richer API in those cases where we can make use of it.

Anyway, again, this is not a huge deal so from my point of view so I'm fine with whichever way @skunert wants to go.

[the remark this answers is not in the document]
Comment (skunert) [reply client/relay-chain-rpc-interface/src/rpc_client.rs]: Lets stick to the tokio bounded channel in this case. Its more defensive here with no real downsides.

File: client/relay-chain-rpc-interface/src/rpc_client.rs (hunk 8 above)
Comment (skunert) [reply client/relay-chain-rpc-interface/src/rpc_client.rs]: Worker took in the streams which required the client. I changed the methods to be static, so now we got rid of the `Option` 👍
